File: 145.go
package leetcode

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func postorderTraversal(root *TreeNode) (ans []int) {
	if root == nil {
		return
	}

	stack := []*TreeNode{root}
	gotoHLVFL := func() {
		// 左侧最高可见节点
		// highest leaf visible from left
		for node := stack[len(stack)-1]; node != nil; node = stack[len(stack)-1] {
			if node.Left != nil {
				// 尽可能向左
				if node.Right != nil {
					stack = append(stack, node.Right)
				}
				stack = append(stack, node.Left)
			} else {
				// 实不得已才向右
				stack = append(stack, node.Right)
			}
		}
		stack = stack[:len(stack)-1]
	}

	for len(stack) > 0 {
		if stack[len(stack)-1].Left != root && stack[len(stack)-1].Right != root {
			// 如果当前栈顶节点不是其父，那就是其右兄
			// 找到其右兄为根的子树中的HLVFL(相当于递归深入其中)
			gotoHLVFL()
		}
		root = stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		ans = append(ans, root.Val)
	}

	return
}

File: 145.py
from typing import List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if root is None:
            return []
        res = []
        stack = [root]
        tem = root
        while stack:
            if stack[-1].left != tem and stack[-1].right != tem:#右兄
                self.goto_height_leaf_visable_from_left(stack)
            tem = stack.pop()
            res.append(tem.val)
        return res

    def goto_height_leaf_visable_from_left(self, s):
        while 1:
            node = s[-1]
            if node is None:
                break
            if node.left is not None:
                if node.right is not None:
                    s.append(node.right)
                s.append(node.left)
            else:
                s.append(node.right)
        s.pop()
    
    def postorderTraversal2(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        s = [root]
        ans = []
        while s:
            node = s.pop()
            ans.append(node.val)
            if node.left:
                s.append(node.left)
            if node.right:
                s.append(node.right)
        return ans[::-1]
